pinot-plugins/pinot-input-format/pinot-csv/src/main/java/org/apache/pinot/plugin/inputformat/csv/CSVDefaultHeaderUtil.java

```diff
@@ -0,0 +1,195 @@
+    HashMap<String, ColumnValueAttributes> csvBodyAttributes = new HashMap<>(numColumns);
+    Iterator<CSVRecord> iterator = records.iterator();
+    if (iterator.hasNext()) {
+      CSVRecord header = iterator.next();
+      for (int i = 0; i < header.size(); i++) {
+        csvHeaderAttributes.put(defaultColumnNames[i], new ColumnValueAttributes(header.get(i), multiValueDelimiter));
+        // Initialize the body attributes with null values at first
+        csvBodyAttributes.put(defaultColumnNames[i], null);
+      }
+    }
+
+    while (iterator.hasNext() && rowCounter < MAX_CSV_SAMPLE_ROWS) {
+      CSVRecord row = iterator.next();
+      // Skip rows that have irregular number of columns
+      if (row.size() != numColumns) {
+        continue;
+      }
+
+      for (int i = 0; i < row.size(); i++) {
+        String fieldName = DEFAULT_CSV_COLUMN_PREFIX + i;
+        String columnValue = row.get(i);
+
+        if (!csvBodyAttributes.containsKey(fieldName) || columnValue == null) {
+          continue;
+        }
+        ColumnValueAttributes csvBodyAttribute = csvBodyAttributes.get(fieldName);
+        ColumnValueAttributes curColumnAttributes = new ColumnValueAttributes(columnValue, multiValueDelimiter);
+        if (csvBodyAttribute == null) {
+          csvBodyAttributes.put(fieldName, curColumnAttributes);
+        } else {
+          // If column type is String and length does not match, remove from consideration
+          if (csvBodyAttribute.getDataType().equals(ColumnValueAttributes.Type.STRING)
+              && csvBodyAttribute.getLength() != curColumnAttributes.getLength()) {
+            // Remove column from consideration
+            csvBodyAttributes.remove(fieldName);
+          } else {
+            // If data type does not match, remove from consideration
+            if (!csvBodyAttribute.getDataType().equals(curColumnAttributes.getDataType())) {
+              csvBodyAttributes.remove(fieldName);
+            }
+          }
+        }
+      }
+    }
+
+    int hasHeader = 0;
+    // Compare first row with the rest of the values
+    for (Map.Entry<String, ColumnValueAttributes> entry : csvBodyAttributes.entrySet()) {
+      ColumnValueAttributes headerAttributes = csvHeaderAttributes.get(entry.getKey());
+
+      if (entry.getValue().getDataType().equals(ColumnValueAttributes.Type.STRING)) {
+        // Compare lengths for String case
+        if (headerAttributes.getLength() != entry.getValue().getLength()) {
+          hasHeader++;
+        } else {
+          hasHeader--;
+        }
+      } else {
+        // Compare types
+        if (!headerAttributes.getDataType().equals(entry.getValue().getDataType())) {
+          hasHeader++;
+        } else {
+          hasHeader--;
+        }
+      }
+    }
+
+    if (hasHeader > 0) {
+      format.withHeader(originalHeader);
+      return null;
+    } else {
+      return StringUtils.join(defaultColumnNames, delimiter);
```

Comment:
So, here we disregard the header even if originalHeader is available ?
Can we add log messages to indicate what actions we are taking.